server/src/main/java/org/elasticsearch/action/synonyms/PutSynonymsAction.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.synonyms;
+
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.ValidateActions;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.StatusToXContentObject;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.synonyms.SynonymSet;
+import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParserConfiguration;
+import org.elasticsearch.xcontent.XContentType;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Objects;
+
+public class PutSynonymsAction extends ActionType<PutSynonymsAction.Response> {
+
+    public static final PutSynonymsAction INSTANCE = new PutSynonymsAction();
+    public static final String NAME = "cluster:admin/synonyms/put";
+
+    public PutSynonymsAction() {
+        super(NAME, Response::new);
+    }
+
+    public static class Request extends ActionRequest {
+        private final String name;
+        private final SynonymSet synonymSet;
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            this.name = in.readString();
+            this.synonymSet = new SynonymSet(in);
+        }
+
+        public Request(String name, BytesReference content, XContentType contentType) throws IOException {
+            this.name = name;
+            this.synonymSet = SynonymSet.fromXContent(XContentHelper.createParser(XContentParserConfiguration.EMPTY, content, contentType));
```

Comment:
Why do we need an extra class `SynonymSet` that just a holder of `List<SynonymRule>`? Can we just read `List<SynonymRule>` here without creating `SynonymSet` object?

Reply:
We might want to add information to it in the future - synonyms format, number of total synonym rules. I think it's more extensible this way.

Reply:
synchronized with @carlosdelest offline, and we decided to keep `SynonymSet` class with an added TODO: to remove it later if not needed.